conn.go

```diff
@@ -1069,3 +1070,57 @@ func (d *connDeadline) unsetConnWriteDeadline() {
 	d.wconn = nil
 	d.mutex.Unlock()
 }
+
+// saslHandshake sends the SASL handshake message
+//
+// See http://kafka.apache.org/protocol.html#The_Messages_SaslHandshake
+func (c *Conn) saslHandshake(request saslHandshakeRequestV1) (saslHandshakeResponseV1, error) {
+	var response saslHandshakeResponseV1
+
+	err := c.writeOperation(
+		func(deadline time.Time, id int32) error {
+			err := c.writeRequest(saslHandshakeRequest, v1, id, request)
+			return err
+		},
+		func(deadline time.Time, size int) error {
+			return expectZeroSize(func() (remain int, err error) {
+				return (&response).readFrom(&c.rbuf, size)
+			}())
+		},
+	)
+	if err != nil {
+		return saslHandshakeResponseV1{}, err
+	}
+	if response.ErrorCode != 0 {
+		return saslHandshakeResponseV1{}, Error(response.ErrorCode)
+	}
+
+	return response, nil
+}
+
+// saslAuthenticate sends the SASL authenticate message
+//
+// See http://kafka.apache.org/protocol.html#The_Messages_SaslAuthenticate
+func (c *Conn) saslAuthenticate(request saslAuthenticateRequestV0) (saslAuthenticateResponseV0, error) {
+	var response saslAuthenticateResponseV0
+
+	err := c.writeOperation(
+		func(deadline time.Time, id int32) error {
+			err := c.writeRequest(saslAuthenticateRequest, v0, id, request)
+			return err
+		},
+		func(deadline time.Time, size int) error {
+			return expectZeroSize(func() (remain int, err error) {
+				return (&response).readFrom(&c.rbuf, size)
+			}())
+		},
+	)
+	if err != nil {
+		return saslAuthenticateResponseV0{}, err
+	}
+	if response.ErrorCode != 0 {
```

Comment:
Consider return some an error if error message is set too?  Should we add the message to kafka.Error (currently just an int)?